c7n/resources/ec2.py

```diff
@@ -1322,6 +1322,139 @@ def process(self, instances):
         return instances
 
 
+@actions.register('propagate-spot-tags')
+class PropagateSpotTags(BaseAction):
+    """Propagate Tags that are set at Spot Request level to EC2 instances.
+
+    :Example:
+
+    .. code-block: yaml
+
+        policies:
+          - name: ec2-spot-instances
+            resource: ec2
+          filters:
+            - State.Name: pending
+            - instanceLifecycle: spot
+          actions:
+            - type: propagate-spot-tags
+              only_tags:
+                - Name
+                - BillingTag
+    """
+
+    schema = type_schema(
+        'propagate-spot-tags',
+        **{'only_tags': {'type': 'array', 'items': {'type': 'string'}}})
+
+    permissions = (
+        'ec2:DescribeInstances',
+        'ec2:DescribeSpotInstanceRequests',
+        'ec2:DescribeTags',
+        'ec2:CreateTags')
+
+    MAX_TAG_COUNT = 50
+
+    def _filter_ec2_with_spot_request_id(self, instances):
+        return [i for i in instances if len(i['SpotInstanceRequestId']) > 0]
+
+    filter_spot_instances = ValueFilter({
+        'key': 'InstanceLifecycle',
+        'value': 'spot'}).validate()
+
+    def process(self, instances):
+        updated_instances = []
+        # We only process spot instances which have a spot request id
+        instances = self.filter_spot_instances.process(
```

Comment:
nice use of composition. minor note, you should probably also set an attribute `annotate = False` on the value filter instance to avoid the extra attribute , its harmless in this case though.